modules/packages/OrderedMap.chpl

```diff
@@ -0,0 +1,746 @@
+    */
+    proc ref this(k: keyType) ref where isDefaultInitializable(valType) {
+      _enter(); defer _leave();
+
+      if !_set.contains((k, nil)) then {
+        var defaultValue: valType;
+        _set.add((k, new shared _valueWrapper(defaultValue)?));
+      } 
+
+      ref e = _set.instance._getReference((k, nil));
+
+      ref result = e[1]!.val;
+      return result;
+    }
+
+    pragma "no doc"
+    proc const this(k: keyType) const
+    where shouldReturnRvalueByValue(valType) && !isNonNilableClass(valType) {
+      _enter(); defer _leave();
+
+      // Could halt
+      var e = _set.instance._getValue((k, nil));
+
+      const result = e[1]!.val;
+      return result;
+    }
+
+    pragma "no doc"
+    proc const this(k: keyType) const ref
+    where shouldReturnRvalueByConstRef(valType) && !isNonNilableClass(valType) {
+      _enter(); defer _leave();
+
+      // Could halt
+      var e = _set.instance._getValue((k, nil));
+
+      const ref result = e[1]!.val;
+      return result;
+    }
+
+    pragma "no doc"
+    proc const this(k: keyType)
+    where isNonNilableClass(valType) {
+      compilerError("Cannot access non-nilable class directly. Use an",
+                    " appropriate accessor method instead.");
+    }
+
+    /* Get a borrowed reference to the element at position `k`.
+     */
+    proc getBorrowed(k: keyType) where isClass(valType) {
+      _enter(); defer _leave();
+
+      // This could halt
+      ref element = _set.instance._getReference((k, nil));
+
+      var result = element[1]!.val.borrow();
+
+      return result;
+    }
+
+    /* Get a reference to the element at position `k`. This method is not
+       available for non-nilable types.
+     */
+    proc getReference(k: keyType) ref
+    where !isNonNilableClass(valType) {
+      _enter(); defer _leave();
+
+      // This could halt
+      ref element = _set.instance._getReference((k, nil));
+
+      ref result = element[1]!.val;
+
+      return result;
+    }
+
+    /*
+      Get a copy of the element stored at position `k`. This method is only
+      available when a orderedMap's `valType` is a non-nilable class.
+    */
+    proc getValue(k: keyType) const {
+      if isOwnedClass(valType) then
+        compilerError('getValue cannot be called when a orderedMap value type ',
+                      'is an owned class, use getBorrowed instead');
+
+      _enter(); defer _leave();
+
+      var result: _eltType;
+      var found: bool;
+      (found, result) = _set.lowerBound((k, nil));
+      if !found || result[0] != k then
+        boundsCheckHalt("orderedMap index " + k:string + " out of bounds");
+      return result[1]!.val;
+    }
+    /*
+      Remove the element at position `k` from the orderedMap and return its value
+    */
+    proc getAndRemove(k: keyType) {
+      _enter(); defer _leave();
+
+      var result: _eltType;
+      var found: bool;
+      (found, result) = _set.lowerBound((k, nil));
+      if !found || result[0] != k then
```

Comment:
Should this use the comparator to compare `result[0]` to `k`?

Reply:
Sounds reasonable

Reply:
Are you going to make that change?

Reply:
Oh. Maybe I've clicked on this by accident. I will make the change

Reply:
Updated